src/thing_physics.c

```diff
@@ -69,12 +68,194 @@ TbBool thing_above_flight_altitude(const struct Thing* thing)
+      if ( x_pos != x_thing )
+      {
+        if ( x_pos <= x_thing )
+          x_thing = ((x_pos - sizexy & 0xFF00) + sizexy + 256);
+        else
+          x_thing = (((sizexy + x_pos) & 0xFF00) - sizexy - 1);
+      }
+      y_thing = thing->mappos.y.val;
+      y_pos = pos->y.val;
+      if ( y_pos != y_thing )
+      {
+        if ( y_pos <= y_thing )
+          y_thing = ((y_pos - sizexy & 0xFF00) + sizexy + 256);
+        else
+          y_thing = (((sizexy + y_pos) & 0xFF00) - sizexy - 1);
+      }
+      pos->x.val = x_thing;
+      pos->y.val = y_thing;
+      break;
+    case SlbBloF_WalledZ:
+      pos->z.val = get_slide_z_coord(thing, pos);
+      break;
+    case SlbBloF_WalledZ|SlbBloF_WalledX:
+      x_thing = thing->mappos.x.val;
+      sizexy = (unsigned short)actual_sizexy_to_nav_sizexy_table[thing->clipbox_size_xy] >> 1;
+      x_pos = pos->x.val;
+      if ( x_pos != x_thing )
+      {
+        if ( x_pos <= x_thing )
+          x_thing = (((x_pos - sizexy) & 0xFF00) + sizexy + 256);
+        else
+          x_thing = (((sizexy + x_pos) & 0xFF00) - sizexy - 1);
+      }
+      pos->x.val = x_thing;
+      pos->z.val = get_slide_z_coord(thing, pos);
+      break;
+    case SlbBloF_WalledZ|SlbBloF_WalledY:
+      y_thing = thing->mappos.y.val;
+      sizexy = (unsigned short)actual_sizexy_to_nav_sizexy_table[thing->clipbox_size_xy] >> 1;
+      y_pos = pos->y.val;
+      if ( y_thing != y_pos )
+      {
+        if ( y_thing >= y_pos )
+          y_thing = ((y_pos - sizexy & 0xFF00) + sizexy + 256);
+        else
+          y_thing = (((y_pos + sizexy) & 0xFF00) - sizexy - 1);
+      }
+      pos->y.val = y_thing;
+      pos->z.val = get_slide_z_coord(thing, pos);
+      break;
+    case SlbBloF_WalledX|SlbBloF_WalledY|SlbBloF_WalledZ:
+      x_thing = thing->mappos.x.val;
+      sizexy = (unsigned short)actual_sizexy_to_nav_sizexy_table[thing->clipbox_size_xy] >> 1;
+      x_pos = pos->x.val;
+      if ( x_pos != x_thing )
+      {
+        if ( x_pos <= x_thing )
+          x_thing = (((x_pos - sizexy) & 0xFF00) + sizexy + 256);
+        else
+          x_thing = (((sizexy + x_pos) & 0xFF00) - sizexy - 1);
+      }
+      y_pos = pos->y.val;
+      y_thing = thing->mappos.y.val;
+      if ( y_pos != y_thing )
+      {
+        if ( y_pos <= y_thing )
+          y_thing = (((y_pos - sizexy) & 0xFF00) + sizexy + 256);
+        else
+          y_thing = (((sizexy + y_pos) & 0xFF00) - sizexy - 1);
+      }
+      pos->x.val = x_thing;
+      pos->y.val = y_thing;
+      pos->z.val = get_slide_z_coord(thing, pos);
+      break;
+    default:
+      return;
+  }
 }
 
-void bounce_thing_off_wall_at(struct Thing *thing, struct Coord3d *pos, long a3)
+void bounce_thing_off_wall_at(struct Thing *thing, struct Coord3d *pos, long blocked_flags)
 {
-    _DK_bounce_thing_off_wall_at(thing, pos, a3); return;
+   // _DK_bounce_thing_off_wall_at(thing, pos, a3); return;
+  short x = (short)thing->veloc_base.x.val;
```

Comment:
These short x, short y, short z lines, shouldn't that be:

```
    struct Coord3d pos2;
    pos2.x.val = 
    pos2.y.val = 
    pos2.z.val = 
```

As seen in project_point_to_wall_on_angle

Reply:
I don't know; that's what's in the original code. Besides, what's the point in having an entire Coord3d structure?

Reply:
I don't know either, but keep in mind you're writing more KeeperFX code, not simply making the decompiled code functional.
In other words, it's not only important that the code functions the same as original, but also that it's somewhat consistent with the practices of the other functions and the function is easy to understand.

That being said, I checked the decompiled code, and these two functions do not seem to match there, so probably not.